design/one-pager-helm-provider.md

```diff
@@ -0,0 +1,400 @@
+            monitoringService: monitoring.googleapis.com/kubernetes
+            networkRef:
+              name: gcp-wpclusters-network
+            location: us-central1
+            locations:
+              - us-central1-a
+          writeConnectionSecretToRef:
+            namespace: crossplane-system
+          reclaimPolicy: Delete
+      patches:
+        - fromFieldPath: "metadata.name"
+          toFieldPath: "metadata.name"
+        - fromFieldPath: "metadata.name"
+          toFieldPath: "spec.writeConnectionSecretToRef.name"
+      connectionDetails:
+        - fromConnectionSecretKey: kubeconfig
+    - base:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
+        spec:
+          forProvider:
+            databaseVersion: MYSQL_5_6
+            region: us-central1
+            settings:
+              tier: db-custom-1-3840
+              dataDiskType: PD_SSD
+              ipConfiguration:
+                ipv4Enabled: true
+                authorizedNetworks:
+                  - value: "0.0.0.0/0"
+          writeConnectionSecretToRef:
+            namespace: crossplane-system
+          providerRef:
+            name: gcp-provider
+          reclaimPolicy: Delete
+      patches:
+        - fromFieldPath: "metadata.uid"
+          toFieldPath: "spec.writeConnectionSecretToRef.name"
+          transforms:
+            - type: string
+              string:
+                fmt: "%s-postgresql"
+        - fromFieldPath: "spec.parameters.storageGB"
+          toFieldPath: "spec.forProvider.settings.dataDiskSizeGb"
+      connectionDetails:
+        - fromConnectionSecretKey: username
+        - fromConnectionSecretKey: password
+        - fromConnectionSecretKey: endpoint
+        - name: port
+          value: "5432"
+    - base:
+        apiVersion: kubernetes.crossplane.io/v1alpha1
+        kind: Provider
+      patches:
+        - fromFieldPath: "metadata.name"
+          toFieldPath: "metadata.name"  
+        - fromFieldPath: "metadata.name"
+          toFieldPath: "spec.credentialsSecretRef.name"       
+    - base:
+        apiVersion: helm.crossplane.io/v1alpha1
+        kind: Release
+        spec:
+          forProvider:
+            chart:
+              name: wordpress
+              repository: https://charts.bitnami.com/bitnami
+            namespace: wordpress
+            values: |
+              mariadb.enabled: false
+              externaldb.enabled: true
+            set:
+            - name: externalDatabase.host
+              valueFrom:
+                secretKeyRef:
+                  key: host
+            - name: externalDatabase.user
+              valueFrom:
+                secretKeyRef:
+                  key: username
+            - name: externalDatabase.password
+              valueFrom:
+                secretKeyRef:
+                  key: password
+            - name: blogName
+          reclaimPolicy: Delete
```

Comment:
Is this something each resource controller needs to implement or `providerRef.name` set properly by composition controllers?

Reply:
You can implement it manually but there is a set of types & functions that you utilize in the controller-runtime. Here is an example from CloudSQLInstance referring to a Network:
* The types used for `*Ref` and `*Selector` fields: https://github.com/crossplane/provider-gcp/blob/master/apis/database/v1beta1/cloudsql_instance_types.go#L366
* Actual resolver code: https://github.com/crossplane/provider-gcp/blob/master/apis/database/v1beta1/referencers.go#L30

The resolver interface `func (mg *CloudSQLInstance) ResolveReferences(ctx context.Context, c client.Reader) error` is something we expect a resource to implement if they'd like their references to be resolved. In this case it's only 1 field, but you can grow this function as much as you'd like to.

In the managed reconciler, you can see that it's called before any of `ExternalClient` functions: https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L520 What the provided ready-made functions do is to check the `Ready` condition of the referenced object and return error if it's not ready so that the reconciler is blocked but you can just copy that function and put your own readiness check as you'd like.

Reply:
I see, thanks for the details!